compiler/optimizer/LoopCanonicalizer.cpp

```diff
@@ -2088,8 +2115,47 @@ void TR_LoopCanonicalizer::eliminateRedundantInductionVariablesFromLoop(TR_Regio
       }
    }
 
+bool TR_LoopCanonicalizer::isTrivialArrayOperation(TR::NodeChecklist &visited,TR::Node *node)
+   {
+   /* While investigating a regression issue we discovered that induction
+      variable elimination opt was causing performance degradation. The
+      opt rewrites an induction variable in terms of another induction
+      variable. This would be ok if both induction variables were equal
+      outside the loop. However, for when they are not, variable
+      elimination leads to creation of an extra tmp variable. In this
+      second case having an extra tmp variable results in performance
+      degradation. Instead of trying to figure out if both induction
+      variables are equal outside the loop we have decided to simply
+      disallow the reduction for loops which are simple enough to be
+      left as is, i.e. the induction variables are left visible for
+      later optimizations. We use isTrivialArrayOperation to recognize
+      such patterns.
+    */
```

Comment:
We should turn this into proper Doxygen documentation, how about the following:

```
/**
 * \brief
 *    Determines whether the node is a trivial (arithmetic and checks only) array operation.
 *
 * \param visited
 *    The list of visited descendants which should not be examined.
 *
 * \param node
 *    The node to examine.
 *
 * \return
 *    \c true if \p node is a trivial array operation; \c false otherwise.
 *
 * \details
 *    This API is used to identify trivial array operations while performing induction variable elimination. In cases
 *    where the array operation is considered trivial there may be a negative trade-off between eliminating an extra
 *    induction variable compared to leaving it present. The typical scenario this can be encountered in is an array
 *    copy operation which may look like this (in pseudocode):
 *
 *    ```
 *    while (j < x) {
 *        a[i++] = b[j++];
 *    }
 *    ```
 *
 *    If induction variables `i` and `j` are not known to be constants outside the loop then eliminating one induction
 *    variable forces us to rewrite the above loop to be:
 *
 *    ```
 *    temp = i - j;
 *    while (j < x) {
 *        a[temp + j] = b[j++]
 *    }
 *    i = temp + j;
 *    ```
 *
 *    i.e. we have to rewrite one induction variable in terms of the other. The latter loop is more complex than the
 *    former because the recalculation of the induction variable now dominates the computation within the loop and
 *    hence it is not a worth while trade-off. Performing such an induction variable elimination can result in degraded
 *    performance which could then be amplified by loop unrolling at later stages of the optimization strategy.
 */
bool TR_LoopCanonicalizer::isTrivialArrayOperation(TR::NodeChecklist &visited, TR::Node *node)
   {
```